spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/caffe/CaffeLoader.scala

```diff
@@ -0,0 +1,378 @@
+    loadCaffe(prototxtPath, modelPath)
+    val parameterTable = model.getParametersTable()
+
+    parameterTable.foreach {
+      case (name: String, params: Table) =>
+        copyParameter(name, params)
+    }
+    model
+  }
+
+  private def copyParameter(name: String, params: Table): Unit = {
+    if (params == null || (!params.contains("weight") && !params.contains("bias"))) return
+    if (!name2LayerV2.contains(name) && !name2LayerV1.contains(name)) {
+      if (matchAll) throw new Exception(s"module $name cannot map a layer in caffe model")
+      logger.info(s"$name uses initialized parameters")
+      return
+    }
+    loadParameters(name, params)
+  }
+
+/**
+ * Load caffe model from prototxt file and binary pre-trained model and converted
+ * to BigDL graph module
+ * Pre-defined module structure is not needed, it will be created dynamically
+ * @return BigDL model and criterion
+ */
+  def createCaffeModel(): (Module[T], ParallelCriterion[T]) = {
+    loadCaffe(prototxtPath, modelPath)
+    val layers = createLayers()
+    val inputs = layers.filter(layer => layer.prevNodes.size == 0).toArray
+    val outputs = layers.filter(layer => layer.nextNodes.size == 0).toArray
+    val module = Graph(inputs, outputs)
+    module.setName(netparam.getName)
+    copyParameters(module)
+    (module, criterions)
+  }
+
+  private def createLayers() : ArrayBuffer[ModuleNode[T]] = {
+    val layers = ArrayBuffer[ModuleNode[T]]()
+    val layersMap = new mutable.HashMap[String, ModuleNode[T]]()
+    val top2LayerMap = new mutable.HashMap[String, String]()
+    val splitLayerMap = new mutable.HashMap[String, ModuleNode[T]]()
+    val allLayers = ArrayBuffer[GeneratedMessage]()
+    if (netparam.getLayersList.size > 0 ) {
+      // filter out those layers from prototxt but also occurs in binary
+      var localMap = new mutable.HashMap[String, Int]()
+      var i = 0
+      netparam.getLayersList.asScala.
+        foreach(layer => {
+          if (!localMap.contains(layer.getName)) {
+            allLayers.append(layer)
+            localMap(layer.getName) = i
+            i += 1
+          } else {
+            allLayers.update(localMap(layer.getName), layer)
+          }
+        })
+    } else {
+      // filter out those layers from prototxt but also occurs in binary
+      var localMap = new mutable.HashMap[String, Int]()
+      var i = 0
+      netparam.getLayerList.asScala.
+        foreach(layer => {
+          if (!localMap.contains(layer.getName)) {
+            allLayers.append(layer)
+            localMap(layer.getName) = i
+            i += 1
+          } else {
+            allLayers.update(localMap(layer.getName), layer)
+          }
+        })
+    }
+    allLayers.foreach(layer => {
+      var name : String = null
+      val topList = new ArrayBuffer[String]()
+      val bottomList = new ArrayBuffer[String]()
+      if (layer.isInstanceOf[LayerParameter]) {
+        val layerParameter = layer.asInstanceOf[LayerParameter]
```

Comment:
It  is better to use pattern matching to replace this part of code without isInstanceOf and asInstanceOf.
For example, something like:
```scala
layer match  {
case  l: LayerParameter => {do something}
case _ => {}
}
```